ldap/servers/plugins/dynamic_lists/dynamic_lists.c

```diff
@@ -0,0 +1,301 @@
+
+    if (slapi_op_internal(pb)) {
+        /* We do not build dynamic content for internal operations*/
+        return rc;
+    }
+
+    /*
+     * Check if the entry has the configured dynamic list objectclass
+     */
+    oc_value = slapi_value_new_string(listConfig.oc);
+    if (slapi_entry_attr_has_syntax_value(list_entry, "objectclass", oc_value) == 0) {
+        /* Entry does not have the groupOfUrls objectclass */
+        goto done;
+    }
+
+    /* The entry could have multiple urls, so we need to loop here */
+    urls = slapi_entry_attr_get_valuearray(list_entry, listConfig.URLAttr);
+    for (size_t i = 0; urls && urls[i]; i++) {
+        Slapi_ValueSet *list_set = NULL;
+        url = slapi_value_get_string(urls[i]);
+        /*
+        * Parse the LDAP URL and validate it
+        */
+        if ((rc = slapi_ldap_url_parse(url, &ludp, 0, &secure)) != 0) {
+            /* Failed to parse the memberUrl attribute */
+            slapi_log_err(SLAPI_LOG_ERR, DYNAMIC_LISTS_PLUGIN_SUBSYSTEM,
+                          "dynamic_lists_pre_entry - %s - failed to parse the LDAP url: %s\n",
+                          slapi_entry_get_dn_const(list_entry), url);
+            continue;
+        }
+
+        if (ludp->lud_filter == NULL) {
+            slapi_log_err(SLAPI_LOG_ERR, DYNAMIC_LISTS_PLUGIN_SUBSYSTEM,
+                          "dynamic_lists_pre_entry - %s - Missing filter in LDAP URL %s\n",
+                          slapi_entry_get_dn_const(list_entry), url);
+            continue;
+        }
+
+        if (ludp->lud_dn == NULL) {
+            slapi_log_err(SLAPI_LOG_ERR, DYNAMIC_LISTS_PLUGIN_SUBSYSTEM,
+                          "dynamic_lists_pre_entry - %s - Missing base dn in LDAP URL %s\n",
+                          slapi_entry_get_dn_const(list_entry), url);
+            continue;
+        }
+
+        if (ludp->lud_attrs != NULL && ludp->lud_attrs[0] != NULL) {
+            /* If an attribute is specified use it */
+            list_attr = ludp->lud_attrs[0];
+        }
+
+        /*
+        * Check if the client has permission to read the list attribute
+        */
+        if (dynamic_lists_check_access(pb, list_entry,
+                                       list_attr ? list_attr : listConfig.attr) == false)
+        {
+            /* Access denied, so we can't proceed */
+            continue;
+        }
+
+        slapi_log_err(SLAPI_LOG_PLUGIN, DYNAMIC_LISTS_PLUGIN_SUBSYSTEM,
+                      "dynamic_lists_pre_entry - %s (url: %s) - Performing internal search\n",
+                      slapi_entry_get_dn_const(list_entry), url);
+
+        /*
+        * Get our DN's groupOfUrls attribute
+        */
+        search_pb = slapi_pblock_new();
+        slapi_search_internal_set_pb(search_pb, ludp->lud_dn, ludp->lud_scope, ludp->lud_filter,
+                                     NULL, 0, NULL, NULL, dynamic_lists_get_plugin_id(), 0);
+        slapi_search_internal_pb(search_pb);
```

Comment:
Also, it always runs an internal search as the plugin identity (effectively `cn=Directory Manager`?) and copies the first value of any attribute named in the LDAP URL straight into the entry being returned to the client. So anybody who can read a dynamic list entry can exfiltrate otherwise protected attributes (e.g. userPassword, homePhone) simply by pointing memberURL at them.
Do I understand the logic correctly or do I miss something?